mlflow/models/signature.py

```diff
@@ -0,0 +1,449 @@
+
+        :return: dictionary representation with input and output shcema represented as json strings.
+        """
+
+        return {
+            "inputs": self.inputs.to_json(),
+            "outputs": self.outputs.to_json() if self.outputs is not None else None
+        }
+
+    @classmethod
+    def from_dict(cls, signature_dict: Dict[str, str]):
+        """
+        Deserialize from dictionary representation.
+
+        :param signature_dict: Dictionary representation of model signature.
+                               Expected dictionary format:
+                               `{'inputs': <json string>, 'outputs': <json string>" }`
+
+        :return: ModelSignature populated with the data form the dictionary.
+        """
+        inputs = Schema.from_json(signature_dict["inputs"])
+        if "outputs" in signature_dict and signature_dict["outputs"] is not None:
+            outputs = Schema.from_json(signature_dict["outputs"])
+            return cls(inputs, outputs)
+        else:
+            return cls(inputs)
+
+    def __eq__(self, other) -> bool:
+        return self.inputs == other.inputs and self.outputs == other.outputs
+
+    def __repr__(self) -> str:
+        return json.dumps({"ModelSignature": self.to_dict()}, indent=2)
+
+
+ModelInputExample = TypeVar('ModelInputExample', pd.DataFrame, np.ndarray, dict, list)
+MlflowModelDataset = TypeVar('MlflowModelDataset', pd.DataFrame, np.ndarray, Dict[str, np.ndarray])
+
+
+def infer_signature(model_input: MlflowModelDataset,
+                    model_output: MlflowModelDataset = None) -> ModelSignature:
+    """
+    Infer an MLflow model signature from the training data (input) and model predictions (output).
+    This method captures the column names and data types from the user data. The signature
+    represents model input and output as dataframes with (optionally) named columns and data type
+    specified as one of types defined in :py:class:`DataType`. This method will raise
+    an exception if the user data contains incompatible types or is not passed in one of the
+    supported formats (containers).
+
+    Supported input:
+
+    container types:
+      - pandas.DataFrame
+      - dictionary of { name -> numpy.ndarray}
+      - numpy.ndarray
+      - pyspark.sql.DataFrame
+
+    element (scalar) types:
+      - those that can be translated to one of :py:class:`mlflow.models.signature.DataType` values.
+
+    NOTE: Multidimensional (>2d) arrays (aka tensors) are not supported at this time.
+
+    :param model_input: Valid input to the model. E.g. (a subset of) training dataset.
+    :param model_output: Valid model output. Model predictions for the (subset of) training dataset.
+    :return: ModelSignature
+    """
+    inputs = _infer_schema(model_input)
+    outputs = _infer_schema(model_output) if model_output is not None else None
+    return ModelSignature(inputs, outputs)
+
+
+def save_example(path: str, input_example: ModelInputExample) -> str:
```

Comment:
Would it make sense to split this file up at all? Right now, it handles a few different concerns -- examples and signature/schema inference seem pretty independent. Also some random stuff about loading DataFrames from JSON. Could have `example.py`, `signature.py`, and `utils.py`, for example, might make it a bit easier to read and understand.

Reply:
makes sense, will do